In [ ]:
! pip install seaborn 

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings


warnings.filterwarnings("ignore")

%matplotlib inline

In [ ]:
! pip install openpyxl

In [5]:
df = pd.read_excel('D:\GUVI_Capstone_Project\Census_data\census_2011 (1).xlsx',header=0)
df.head()

District code         State name District name  Population      Male  \
0              1  JAMMU AND KASHMIR       Kupwara    870354.0  474190.0   
1              2  JAMMU AND KASHMIR        Badgam    753745.0       NaN   
2              3  JAMMU AND KASHMIR   Leh(Ladakh)    133487.0   78971.0   
3              4  JAMMU AND KASHMIR        Kargil    140802.0       NaN   
4              5  JAMMU AND KASHMIR         Punch         NaN  251899.0   

     Female  Literate  Male_Literate  Female_Literate      SC  ...  \
0  396164.0  439654.0       282823.0         156831.0  1048.0  ...   
1  355704.0  335649.0       207741.0         127908.0     NaN  ...   
2   54516.0   93770.0        62834.0          30936.0   488.0  ...   
3   63017.0       NaN        56301.0          29935.0    18.0  ...   
4  224936.0  261724.0       163333.0          98391.0   556.0  ...   

   Power_Parity_Rs_90000_150000  Power_Parity_Rs_45000_150000  \
0                          94.0                         588.0   
1                         126.0                         562.0   
2                          46.0                         122.0   
3                          27.0                         114.0   
4                          78.0                         346.0   

   Power_Parity_Rs_150000_240000  Power_Parity_Rs_240000_330000  \
0                           71.0                          101.0   
1                           72.0                           89.0   
2                           15.0                           22.0   
3                           12.0                           18.0   
4                           35.0                           50.0   

   Power_Parity_Rs_150000_330000  Power_Parity_Rs_330000_425000  \
0                          172.0                           74.0   
1                          161.0                           96.0   
2                            NaN                           20.0   
3                           30.0                           19.0   
4                           85.0                           59.0   

   Power_Parity_Rs_425000_545000  Power_Parity_Rs_330000_545000  \
0                           10.0                           84.0   
1                           28.0                          124.0   
2                            NaN                            NaN   
3                            3.0                           22.0   
4                            8.0                           67.0   

   Power_Parity_Above_Rs_545000  Total_Power_Parity  
0                          15.0              1119.0  
1                          18.0              1066.0  
2                          17.0               242.0  
3                           7.0               214.0  
4                          12.0               629.0  

[5 rows x 118 columns]

#### Shape of the dataset:

In [7]:
df.shape

(640, 118)

The dataset contains 640 rows and 118 columns.

#### Summary of the dataset :

In [8]:
# Displaying summary for a dataframe

df.describe()

District code    Population          Male        Female      Literate  \
count     640.000000  6.100000e+02  6.100000e+02  6.070000e+02  6.040000e+02   
mean      320.500000  1.852620e+06  9.685726e+05  9.139882e+05  1.190550e+06   
std       184.896367  1.435578e+06  7.926455e+05  7.479293e+05  1.050630e+06   
min         1.000000  8.004000e+03  4.414000e+03  3.590000e+03  4.436000e+03   
25%       160.750000  8.148625e+05  4.180382e+05  3.986990e+05  4.796545e+05   
50%       320.500000  1.557367e+06  7.939620e+05  7.556020e+05  9.793575e+05   
75%       480.250000  2.565248e+06  1.318615e+06  1.253832e+06  1.602260e+06   
max       640.000000  1.000978e+07  5.865078e+06  5.195070e+06  8.227161e+06   

       Male_Literate  Female_Literate            SC       Male_SC  \
count   6.090000e+02     6.130000e+02  6.050000e+02  6.170000e+02   
mean    6.739088e+05     5.129481e+05  3.177854e+05  1.624509e+05   
std     5.683384e+05     4.801441e+05  3.172010e+05  1.625607e+05   
min     2.614000e+03     1.822000e+03  0.000000e+00  0.000000e+00   
25%     2.695820e+05     2.009450e+05  8.799100e+04  4.251300e+04   
50%     5.507160e+05     4.048610e+05  2.463370e+05  1.252140e+05   
75%     9.203140e+05     6.568420e+05  4.550620e+05  2.284310e+05   
max     4.591396e+06     3.635765e+06  2.464032e+06  1.266504e+06   

          Female_SC  ...  Power_Parity_Rs_90000_150000  \
count  6.080000e+02  ...                    607.000000   
mean   1.544700e+05  ...                    792.028007   
std    1.537418e+05  ...                   1059.010710   
min    0.000000e+00  ...                      0.000000   
25%    4.411825e+04  ...                    236.000000   
50%    1.181440e+05  ...                    507.000000   
75%    2.163095e+05  ...                    939.000000   
max    1.197528e+06  ...                  10334.000000   

       Power_Parity_Rs_45000_150000  Power_Parity_Rs_150000_240000  \
count                    610.000000                     611.000000   
mean                    1676.009836                     288.960720   
std                     1669.641260                     641.765892   
min                        0.000000                       0.000000   
25%                      588.000000                      59.000000   
50%                     1196.000000                     144.000000   
75%                     2224.250000                     293.000000   
max                    13483.000000                   10835.000000   

       Power_Parity_Rs_240000_330000  Power_Parity_Rs_150000_330000  \
count                     605.000000                     616.000000   
mean                      213.919008                     513.099026   
std                       370.466703                     985.702797   
min                         0.000000                       0.000000   
25%                        23.000000                      93.500000   
50%                       107.000000                     274.000000   
75%                       253.000000                     562.250000   
max                      3595.000000                   14430.000000   

       Power_Parity_Rs_330000_425000  Power_Parity_Rs_425000_545000  \
count                     607.000000                     610.000000   
mean                      197.375618                     264.191803   
std                       433.595214                     599.182627   
min                         0.000000                       0.000000   
25%                        19.000000                      21.000000   
50%                        84.000000                      84.000000   
75%                       216.000000                     293.000000   
max                      5027.000000                    7597.000000   

       Power_Parity_Rs_330000_545000  Power_Parity_Above_Rs_545000  \
count                     617.000000                    610.000000   
mean                      463.019449                    283.934426   
std                 

### Check Datatypes in the dataset

In [9]:
# Check Null and Dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 640 entries, 0 to 639
Columns: 118 entries, District code to Total_Power_Parity
dtypes: float64(115), int64(1), object(2)
memory usage: 590.1+ KB


### Check for all columns

In [6]:
df.columns

Index(['District code', 'State name', 'District name', 'Population', 'Male',
       'Female', 'Literate', 'Male_Literate', 'Female_Literate', 'SC',
       ...
       'Power_Parity_Rs_90000_150000', 'Power_Parity_Rs_45000_150000',
       'Power_Parity_Rs_150000_240000', 'Power_Parity_Rs_240000_330000',
       'Power_Parity_Rs_150000_330000', 'Power_Parity_Rs_330000_425000',
       'Power_Parity_Rs_425000_545000', 'Power_Parity_Rs_330000_545000',
       'Power_Parity_Above_Rs_545000', 'Total_Power_Parity'],
      dtype='object', length=118)

## Task 1 : Remane the Column names

In [7]:
def rename_columns(df):
    column_mapping = {
        'State name': 'State/UT',
        'District name': 'District',
        'Male_Literate': 'Literate_Male',
        'Female_Literate': 'Literate_Female',
        'Rural_Households': 'Households_Rural',
        'Urban_ Households': 'Households_Urban',
        'Age_Group_0_29': 'Young_and_Adult',
        'Age_Group_30_49': 'Middle_Aged',
        'Age_Group_50': 'Senior_Citizen',
        'Age not stated': 'Age_Not_Stated'
    }
    print(f"Original columns: {df.columns.tolist()}")
    df.rename(columns=column_mapping, inplace=True)
    print(f"Renamed columns: {df.columns.tolist()}")
    return df

In [8]:
df = rename_columns(df)

Original columns: ['District code', 'State name', 'District name', 'Population', 'Male', 'Female', 'Literate', 'Male_Literate', 'Female_Literate', 'SC', 'Male_SC', 'Female_SC', 'ST', 'Male_ST', 'Female_ST', 'Workers', 'Male_Workers', 'Female_Workers', 'Main_Workers', 'Marginal_Workers', 'Non_Workers', 'Cultivator_Workers', 'Agricultural_Workers', 'Household_Workers', 'Other_Workers', 'Hindus', 'Muslims', 'Christians', 'Sikhs', 'Buddhists', 'Jains', 'Others_Religions', 'Religion_Not_Stated', 'LPG_or_PNG_Households', 'Housholds_with_Electric_Lighting', 'Households_with_Internet', 'Households_with_Computer', 'Rural_Households', 'Urban_Households', 'Households', 'Below_Primary_Education', 'Primary_Education', 'Middle_Education', 'Secondary_Education', 'Higher_Education', 'Graduate_Education', 'Other_Education', 'Literate_Education', 'Illiterate_Education', 'Total_Education', 'Age_Group_0_29', 'Age_Group_30_49', 'Age_Group_50', 'Age not stated', 'Households_with_Bicycle', 'Households_with_C

In [9]:
df.head()

District code           State/UT     District  Population      Male  \
0              1  JAMMU AND KASHMIR      Kupwara    870354.0  474190.0   
1              2  JAMMU AND KASHMIR       Badgam    753745.0       NaN   
2              3  JAMMU AND KASHMIR  Leh(Ladakh)    133487.0   78971.0   
3              4  JAMMU AND KASHMIR       Kargil    140802.0       NaN   
4              5  JAMMU AND KASHMIR        Punch         NaN  251899.0   

     Female  Literate  Literate_Male  Literate_Female      SC  ...  \
0  396164.0  439654.0       282823.0         156831.0  1048.0  ...   
1  355704.0  335649.0       207741.0         127908.0     NaN  ...   
2   54516.0   93770.0        62834.0          30936.0   488.0  ...   
3   63017.0       NaN        56301.0          29935.0    18.0  ...   
4  224936.0  261724.0       163333.0          98391.0   556.0  ...   

   Power_Parity_Rs_90000_150000  Power_Parity_Rs_45000_150000  \
0                          94.0                         588.0   
1                         126.0                         562.0   
2                          46.0                         122.0   
3                          27.0                         114.0   
4                          78.0                         346.0   

   Power_Parity_Rs_150000_240000  Power_Parity_Rs_240000_330000  \
0                           71.0                          101.0   
1                           72.0                           89.0   
2                           15.0                           22.0   
3                           12.0                           18.0   
4                           35.0                           50.0   

   Power_Parity_Rs_150000_330000  Power_Parity_Rs_330000_425000  \
0                          172.0                           74.0   
1                          161.0                           96.0   
2                            NaN                           20.0   
3                           30.0                           19.0   
4                           85.0                           59.0   

   Power_Parity_Rs_425000_545000  Power_Parity_Rs_330000_545000  \
0                           10.0                           84.0   
1                           28.0                          124.0   
2                            NaN                            NaN   
3                            3.0                           22.0   
4                            8.0                           67.0   

   Power_Parity_Above_Rs_545000  Total_Power_Parity  
0                          15.0              1119.0  
1                          18.0              1066.0  
2                          17.0               242.0  
3                           7.0               214.0  
4                          12.0               629.0  

[5 rows x 118 columns]

## Task 2: Rename State/UT Names

In [10]:
def rename_state_ut_names(df):
    if 'State/UT' in df.columns:
        df['State/UT'] = df['State/UT'].apply(lambda x: ' '.join([word.capitalize() if word != 'and' else word.lower() for word in x.split()]))
        print("State/UT names standardized successfully")
    else:
        print("Error: 'State/UT' column not found in DataFrame")
    return df

In [11]:
df = rename_state_ut_names(df)

State/UT names standardized successfully


In [12]:
df.head(5)

District code           State/UT     District  Population      Male  \
0              1  Jammu And Kashmir      Kupwara    870354.0  474190.0   
1              2  Jammu And Kashmir       Badgam    753745.0       NaN   
2              3  Jammu And Kashmir  Leh(Ladakh)    133487.0   78971.0   
3              4  Jammu And Kashmir       Kargil    140802.0       NaN   
4              5  Jammu And Kashmir        Punch         NaN  251899.0   

     Female  Literate  Literate_Male  Literate_Female      SC  ...  \
0  396164.0  439654.0       282823.0         156831.0  1048.0  ...   
1  355704.0  335649.0       207741.0         127908.0     NaN  ...   
2   54516.0   93770.0        62834.0          30936.0   488.0  ...   
3   63017.0       NaN        56301.0          29935.0    18.0  ...   
4  224936.0  261724.0       163333.0          98391.0   556.0  ...   

   Power_Parity_Rs_90000_150000  Power_Parity_Rs_45000_150000  \
0                          94.0                         588.0   
1                         126.0                         562.0   
2                          46.0                         122.0   
3                          27.0                         114.0   
4                          78.0                         346.0   

   Power_Parity_Rs_150000_240000  Power_Parity_Rs_240000_330000  \
0                           71.0                          101.0   
1                           72.0                           89.0   
2                           15.0                           22.0   
3                           12.0                           18.0   
4                           35.0                           50.0   

   Power_Parity_Rs_150000_330000  Power_Parity_Rs_330000_425000  \
0                          172.0                           74.0   
1                          161.0                           96.0   
2                            NaN                           20.0   
3                           30.0                           19.0   
4                           85.0                           59.0   

   Power_Parity_Rs_425000_545000  Power_Parity_Rs_330000_545000  \
0                           10.0                           84.0   
1                           28.0                          124.0   
2                            NaN                            NaN   
3                            3.0                           22.0   
4                            8.0                           67.0   

   Power_Parity_Above_Rs_545000  Total_Power_Parity  
0                          15.0              1119.0  
1                          18.0              1066.0  
2                          17.0               242.0  
3                           7.0               214.0  
4                          12.0               629.0  

[5 rows x 118 columns]

### Define numerical & Categorical columns

In [13]:
numeric_features =[feature for feature in df.columns if df[feature].dtype != 'O']
categorical_features = [feature for feature in df.columns if df[feature].dtype == 'O']

#Print columns
print('We have {} numerical features : {}'.format(len(numeric_features),numeric_features))
print('We have {} categorical features : {}'.format(len(categorical_features),categorical_features))

We have 116 numerical features : ['District code', 'Population', 'Male', 'Female', 'Literate', 'Literate_Male', 'Literate_Female', 'SC', 'Male_SC', 'Female_SC', 'ST', 'Male_ST', 'Female_ST', 'Workers', 'Male_Workers', 'Female_Workers', 'Main_Workers', 'Marginal_Workers', 'Non_Workers', 'Cultivator_Workers', 'Agricultural_Workers', 'Household_Workers', 'Other_Workers', 'Hindus', 'Muslims', 'Christians', 'Sikhs', 'Buddhists', 'Jains', 'Others_Religions', 'Religion_Not_Stated', 'LPG_or_PNG_Households', 'Housholds_with_Electric_Lighting', 'Households_with_Internet', 'Households_with_Computer', 'Households_Rural', 'Urban_Households', 'Households', 'Below_Primary_Education', 'Primary_Education', 'Middle_Education', 'Secondary_Education', 'Higher_Education', 'Graduate_Education', 'Other_Education', 'Literate_Education', 'Illiterate_Education', 'Total_Education', 'Young_and_Adult', 'Middle_Aged', 'Senior_Citizen', 'Age_Not_Stated', 'Households_with_Bicycle', 'Households_with_Car_Jeep_Van', 'Ho

In [14]:
df[numeric_features].head(5)

District code  Population      Male    Female  Literate  Literate_Male  \
0              1    870354.0  474190.0  396164.0  439654.0       282823.0   
1              2    753745.0       NaN  355704.0  335649.0       207741.0   
2              3    133487.0   78971.0   54516.0   93770.0        62834.0   
3              4    140802.0       NaN   63017.0       NaN        56301.0   
4              5         NaN  251899.0  224936.0  261724.0       163333.0   

   Literate_Female      SC  Male_SC  Female_SC  ...  \
0         156831.0  1048.0   1046.0        2.0  ...   
1         127908.0     NaN    343.0       25.0  ...   
2          30936.0   488.0    444.0       44.0  ...   
3          29935.0    18.0     12.0        6.0  ...   
4          98391.0   556.0    406.0        NaN  ...   

   Power_Parity_Rs_90000_150000  Power_Parity_Rs_45000_150000  \
0                          94.0                         588.0   
1                         126.0                         562.0   
2                          46.0                         122.0   
3                          27.0                         114.0   
4                          78.0                         346.0   

   Power_Parity_Rs_150000_240000  Power_Parity_Rs_240000_330000  \
0                           71.0                          101.0   
1                           72.0                           89.0   
2                           15.0                           22.0   
3                           12.0                           18.0   
4                           35.0                           50.0   

   Power_Parity_Rs_150000_330000  Power_Parity_Rs_330000_425000  \
0                          172.0                           74.0   
1                          161.0                           96.0   
2                            NaN                           20.0   
3                           30.0                           19.0   
4                           85.0                           59.0   

   Power_Parity_Rs_425000_545000  Power_Parity_Rs_330000_545000  \
0                           10.0                           84.0   
1                           28.0                          124.0   
2                            NaN                            NaN   
3                            3.0                           22.0   
4                            8.0                           67.0   

   Power_Parity_Above_Rs_545000  Total_Power_Parity  
0                          15.0              1119.0  
1                          18.0              1066.0  
2                          17.0               242.0  
3                           7.0               214.0  
4                          12.0               629.0  

[5 rows x 116 columns]

In [15]:
df[categorical_features].head(5)

State/UT     District
0  Jammu And Kashmir      Kupwara
1  Jammu And Kashmir       Badgam
2  Jammu And Kashmir  Leh(Ladakh)
3  Jammu And Kashmir       Kargil
4  Jammu And Kashmir        Punch

## Task -3 New State/UT formation with encoding handling

- In 2014 Telangana was formed after it split from Andhra Pradesh, The districts that were included in Telangana are stored in Data/Telangana.txt . Read the text file and Rename the State/UT From “Andhra Pradesh” to “Telangana” for the given districts.

- In 2019 Ladakh was formed after it split from Jammu and Kashmir, which included the districts Leh and Kargil.  Rename the State/UT From “Jammu and Kashmir” to “Ladakh” for the given districts.

In [ ]:
! pip install python-docx

In [17]:
import docx

def getText(filename):
    doc = docx.Document(filename)
    fullText = []
    for para in doc.paragraphs:
        fullText.append(para.text)
    return fullText


telanganaDistric = getText('D:\GUVI_Capstone_Project\Census_data\Telangana (1).docx')

print(telanganaDistric)

['Adilabad', 'Nizamabad', 'Karimnagar', 'Medak', 'Hyderabad', 'Rangareddy', 'Mahbubnagar', 'Nalgonda', 'Warangal', 'Khammam']


In [18]:
def setTelangana():
    for  i in telanganaDistric :
        df.loc[ df['District'] == i, 'State/UT'] ='Telangana'

setTelangana()

ladakhDistric= [
    "Leh(Ladakh)",
    "Kargil"
]

def setLadakh():
    for  i in ladakhDistric :
        df.loc[ df['District'] == i, 'State/UT'] ='Ladakh'


setLadakh()

df[df['State/UT'].str.contains('Ladakh')]

District code State/UT     District  Population     Male   Female  \
2              3   Ladakh  Leh(Ladakh)    133487.0  78971.0  54516.0   
3              4   Ladakh       Kargil    140802.0      NaN  63017.0   

   Literate  Literate_Male  Literate_Female     SC  ...  \
2   93770.0        62834.0          30936.0  488.0  ...   
3       NaN        56301.0          29935.0   18.0  ...   

   Power_Parity_Rs_90000_150000  Power_Parity_Rs_45000_150000  \
2                          46.0                         122.0   
3                          27.0                         114.0   

   Power_Parity_Rs_150000_240000  Power_Parity_Rs_240000_330000  \
2                           15.0                           22.0   
3                           12.0                           18.0   

   Power_Parity_Rs_150000_330000  Power_Parity_Rs_330000_425000  \
2                            NaN                           20.0   
3                           30.0                           19.0   

   Power_Parity_Rs_425000_545000  Power_Parity_Rs_330000_545000  \
2                            NaN                            NaN   
3                            3.0                           22.0   

   Power_Parity_Above_Rs_545000  Total_Power_Parity  
2                          17.0               242.0  
3                           7.0               214.0  

[2 rows x 118 columns]

In [19]:
print(((df.isnull() | df.isna()).sum() * 100 / df.index.size).round(2))

District code                    0.00
State/UT                         0.00
District                         0.00
Population                       4.69
Male                             4.69
                                 ... 
Power_Parity_Rs_330000_425000    5.16
Power_Parity_Rs_425000_545000    4.69
Power_Parity_Rs_330000_545000    3.59
Power_Parity_Above_Rs_545000     4.69
Total_Power_Parity               5.00
Length: 118, dtype: float64


## Task 4 : Find and Process Missing Data

In [20]:
df['Population'].fillna(value=df['Male']+df['Female'],inplace=True)

df['Female'].fillna(value=df['Population'] - df['Male'], inplace=True)

df['Male'].fillna(value=df['Population'] - df['Female'], inplace=True)

df['Literate'].fillna(value=df['Literate_Male']+df['Literate_Female'],inplace=True)


df['Literate_Male'].fillna(value=df['Literate'] - df['Literate_Female'], inplace=True)

df['Literate_Female'].fillna(value=df['Literate'] - df['Literate_Male'], inplace=True)

df['ST'].fillna(value=df['Male_ST']+df['Female_ST'],inplace=True)

df['Male_ST'].fillna(value=df['ST'] -df['Female_ST'], inplace=True)

df['Female_ST'].fillna(value=df['ST'] -df['Male_ST'], inplace=True)

df['Workers'].fillna(value=df['Male_Workers']+df['Female_Workers'],inplace=True)

df['Male_Workers'].fillna(value=df['Workers'] -df['Female_Workers'], inplace=True)

df['Female_Workers'].fillna(value=df['Workers'] -df['Male_Workers'], inplace=True)

In [21]:
print(((df.isnull() | df.isna()).sum() * 100 / df.index.size).round(2))

District code                    0.00
State/UT                         0.00
District                         0.00
Population                       0.16
Male                             0.16
                                 ... 
Power_Parity_Rs_330000_425000    5.16
Power_Parity_Rs_425000_545000    4.69
Power_Parity_Rs_330000_545000    3.59
Power_Parity_Above_Rs_545000     4.69
Total_Power_Parity               5.00
Length: 118, dtype: float64


In [22]:
# Replace missing data in each column with the mean of that column
df_filled = df.fillna(df.mean(numeric_only=True))

# Verify that missing values have been filled
missing_values_after = df_filled.isnull().sum()

missing_values_after


District code                    0
State/UT                         0
District                         0
Population                       0
Male                             0
                                ..
Power_Parity_Rs_330000_425000    0
Power_Parity_Rs_425000_545000    0
Power_Parity_Rs_330000_545000    0
Power_Parity_Above_Rs_545000     0
Total_Power_Parity               0
Length: 118, dtype: int64

### Replace missing data in each column with the mean of that column

In [23]:

for column in df.columns:
    if df[column].dtype in ['float64', 'int64']:  # Only calculate the mean for numeric columns
        mean_value = df[column].mean()
        df[column].fillna(mean_value, inplace=True)

# Verify that missing values have been filled
missing_values_after = df.isnull().sum()

missing_values_after

District code                    0
State/UT                         0
District                         0
Population                       0
Male                             0
                                ..
Power_Parity_Rs_330000_425000    0
Power_Parity_Rs_425000_545000    0
Power_Parity_Rs_330000_545000    0
Power_Parity_Above_Rs_545000     0
Total_Power_Parity               0
Length: 118, dtype: int64

In [24]:
# Calculate the mean value of the 'Male' column, excluding NaNs
mean_male = df['Male'].mean()

# Replace missing values in the 'Male' column with the mean value
df['Male'].fillna(mean_male, inplace=True)

# Verify that missing values have been filled in the 'Male' column
missing_values_male_after = df['Male'].isnull().sum()

missing_values_male_after

0

In [25]:
# Calculate the mean value of the 'Female' column, excluding NaNs
mean_female = df['Female'].mean()

# Replace missing values in the 'Female' column with the mean value
df['Female'].fillna(mean_female, inplace=True)

# Verify that missing values have been filled in the 'female' column
missing_values_male_after = df['Female'].isnull().sum()

missing_values_male_after


0

In [26]:
df

District code                     State/UT                   District  \
0                1            Jammu And Kashmir                    Kupwara   
1                2            Jammu And Kashmir                     Badgam   
2                3                       Ladakh                Leh(Ladakh)   
3                4                       Ladakh                     Kargil   
4                5            Jammu And Kashmir                      Punch   
..             ...                          ...                        ...   
635            636                  Pondicherry                       Mahe   
636            637                  Pondicherry                   Karaikal   
637            638  Andaman And Nicobar Islands                   Nicobars   
638            639  Andaman And Nicobar Islands  North  AND Middle Andaman   
639            640  Andaman And Nicobar Islands              South Andaman   

     Population      Male    Female  Literate  Literate_Male  Literate_Female  \
0      870354.0  474190.0  396164.0  439654.0       282823.0         156831.0   
1      753745.0  398041.0  355704.0  335649.0       207741.0         127908.0   
2      133487.0   78971.0   54516.0   93770.0        62834.0          30936.0   
3      140802.0   77785.0   63017.0   86236.0        56301.0          29935.0   
4      476835.0  251899.0  224936.0  261724.0       163333.0          98391.0   
..          ...       ...       ...       ...            ...              ...   
635     41816.0   19143.0   22673.0   36470.0        16610.0          19860.0   
636    200222.0   97809.0  102413.0  154916.0        79903.0          75013.0   
637     36842.0   20727.0   16115.0   25332.0        15397.0           9935.0   
638    105597.0   54861.0   50736.0   78683.0        43186.0          35497.0   
639    238142.0  127283.0  110859.0  190266.0       105794.0          84472.0   

                SC  ...  Power_Parity_Rs_90000_150000  \
0      1048.000000  ...                          94.0   
1    317785.386777  ...                         126.0   
2       488.000000  ...                          46.0   
3        18.000000  ...                          27.0   
4       556.000000  ...                          78.0   
..             ...  ...                           ...   
635     144.000000  ...                        2316.0   
636   35348.000000  ...                        1063.0   
637       0.000000  ...                         685.0   
638       0.000000  ...                         685.0   
639       0.000000  ...                        1371.0   

     Power_Parity_Rs_45000_150000  Power_Parity_Rs_150000_240000  \
0                           588.0                           71.0   
1                           562.0                           72.0   
2                           122.0                           15.0   
3                           114.0                           12.0   
4                           346.0                           35.0   
..                            ...                            ...   
635                        4309.0                         1370.0   
636                        2408.0                          665.0   
637                        1895.0                          212.0   
638                        1895.0                          212.0   
639                        3020.0                          649.0   

     Power_Parity_Rs_240000_330000  Power_Parity_Rs_150000_330000  \
0                       101.000000                     172.000000   
1                        89.000000                     161.000000   
2                        22.000000                     513.099026   
3                        18.000000                      30.000000   
4                        50.000000                      85.000000   
..                             ...                            ...   
635                     838.000000                    2208.000000   
636                     340.000000       

### Resolve all the Nan value & missing value issue 

## Task 5 : Save The Data to MongoDB

In [ ]:
!python -m pip install "pymongo[srv]"

In [27]:
from pymongo.mongo_client import MongoClient
from pymongo.server_api import ServerApi

uri = "mongodb+srv://sashikantaraja1966:12345@cluster0.ndy0rxz.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0"

# Create a new client and connect to the server
client = MongoClient(uri, server_api=ServerApi('1'))

# Send a ping to confirm a successful connection
try:
    client.admin.command('ping')
    print("Pinged your deployment. You successfully connected to MongoDB!")
except Exception as e:
    print(e)

Pinged your deployment. You successfully connected to MongoDB!


In [28]:
def insertToMongodb():
    if client['test'].list_collection_names != "" :
        myCollection = client['test']
        if myCollection['census'].find_one() != "":
            myCollection['census'].insert_many(df.to_dict("records"))
        else:
            print('No Collection inside DataBase')
            myCollection = client['census']
            myCollection['census'].insert_many(df.to_dict("records"))
    else:
        print('No DataBase')
        dp = client['test']
        myCollection = client['census']
        myCollection['census'].insert_many(df.to_dict("records"))
        print('The insertion successfull!')

insertToMongodb()

## Task 6: Database connection to Mysql & Upload the data to Mysql from MongoDB

In [29]:
%pip install mysql-connector-python

Note: you may need to restart the kernel to use updated packages.


In [29]:
import mysql.connector
mydb = mysql.connector.connect(
    host ="localhost",
    user='root',
    password ="",
    connection_timeout=300
)
print(mydb)
mycursor = mydb.cursor(buffered=True)


OperationalError: 2013 (HY000): Lost connection to MySQL server at 'reading initial communication packet', system error: 0

In [50]:
def creatdb(df):
    mycursor.execute('CREATE DATABASE IF NOT EXISTS census_db')

    listOfColumns = list()

    for i in df.columns:
        if i == 'District code':
            listOfColumns.append('District_Code INT PRIMARY KEY')
        elif i == "State/UT":
            listOfColumns.append('State_UT VARCHAR(50)')
        elif i == "District":
            listOfColumns.append('District VARCHAR(50)')
        elif i == "Households_with_TV_Computer_Laptop_Telephone_mobile_phone_and_Scooter_Car":
            listOfColumns.append("Households_TV_Computer_Laptop_Telephone_mobile_phone_Scooter_Car INT")
        elif i == "Type_of_latrine_facility_Night_soil_disposed_into_open_drain_Households":
            listOfColumns.append("Type_of_latrine_facility_Night_soil_disposed_into_open_drain INT")
        elif i == "Type_of_latrine_facility_Flush_pour_flush_latrine_connected_to_other_system_Households":
            listOfColumns.append("Type_of_latrine_Flush_pour_connected_to_other_system_Households INT")
        elif i == "Not_having_latrine_facility_within_the_premises_Alternative_source_Open_Households":
            listOfColumns.append("Not_having_latrine_within_premises_Other_source_Open_Households INT")
        elif i == "Main_source_of_drinking_water_Handpump_Tubewell_Borewell_Households":
            listOfColumns.append("Source_of_drinking_water_Handpump_Tubewell_Borewell_Households INT")
        elif i == "Main_source_of_drinking_water_Other_sources_Spring_River_Canal_Tank_Pond_Lake_Other_sources__Households":
            listOfColumns.append("Drinking_water_Spring_River_Canal_Tank_Pond_Lake_Other_Household INT")
        else:
            listOfColumns.append(f'{i} INT')

    mycursor.execute(f'CREATE TABLE census_db.census_2011 ({", ".join(listOfColumns)})')
    mydb.commit()

creatdb(df)

OperationalError: 2013 (HY000): Lost connection to MySQL server during query

In [32]:
def get_columns(table_name):
    mycursor.execute(f"DESCRIBE {table_name}")
    table_columns = []
    #print(mycursor)
    for i in mycursor:
        table_columns.append(i[0])

    return tuple(table_columns)

table_columns = get_columns('Census_db.census_2011')

print(table_columns)

('District_Code', 'State_UT', 'District', 'Population', 'Male', 'Female', 'Literate', 'Literate_Male', 'Literate_Female', 'SC', 'Male_SC', 'Female_SC', 'ST', 'Male_ST', 'Female_ST', 'Workers', 'Male_Workers', 'Female_Workers', 'Main_Workers', 'Marginal_Workers', 'Non_Workers', 'Cultivator_Workers', 'Agricultural_Workers', 'Household_Workers', 'Other_Workers', 'Hindus', 'Muslims', 'Christians', 'Sikhs', 'Buddhists', 'Jains', 'Others_Religions', 'Religion_Not_Stated', 'LPG_or_PNG_Households', 'Housholds_with_Electric_Lighting', 'Households_with_Internet', 'Households_with_Computer', 'Households_Rural', 'Urban_Households', 'Households', 'Below_Primary_Education', 'Primary_Education', 'Middle_Education', 'Secondary_Education', 'Higher_Education', 'Graduate_Education', 'Other_Education', 'Literate_Education', 'Illiterate_Education', 'Total_Education', 'Young_and_Adult', 'Middle_Aged', 'Senior_Citizen', 'Age_Not_Stated', 'Households_with_Bicycle', 'Households_with_Car_Jeep_Van', 'Households_

In [33]:
def dbinsert_data(clients, df, table_columns):
    placeholders = ", ".join(["%s"] * len(table_columns))
    columns = ", ".join(table_columns)

    data = []

    for i in range(1,len(df)+1):
        for j in clients.find({'District code':i}):
            data.append(tuple(j.values())[1:])

    try :
        query = f'INSERT INTO Census_DB.Census_2011 ({columns}) VALUES ({placeholders})'
        mycursor.executemany(query, data)
    except Exception as e:
        print(e)

    mydb.commit()
my_dbs = client['test']
mycolle = my_dbs['census']
dbinsert_data(mycolle, df, table_columns)

1054 (42S22): Unknown column 'nan' in 'field list'
